flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/GlobalStatisticsAggregatorTracker.java

```diff
@@ -0,0 +1,142 @@
+
+import java.io.Serializable;
+import org.apache.flink.annotation.Internal;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * GlobalStatisticsAggregator is used by {@link DataStatisticsCoordinator} to collect {@link
+ * DataStatistics} from {@link DataStatisticsOperator} subtasks for specific checkpoint. It stores
+ * the merged {@link DataStatistics} result and uses set to keep a record of all reported subtasks.
+ */
+@Internal
+class GlobalStatisticsAggregatorTracker<K> implements Serializable {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(GlobalStatisticsAggregatorTracker.class);
+  private static final double EXPECTED_DATA_STATISTICS_RECEIVED_PERCENTAGE = 90;
+  private final transient DataStatisticsFactory<K> statisticsFactory;
+  private final transient int parallelism;
+  private transient volatile GlobalStatisticsAggregator<K> inProgressAggregator;
+  private volatile GlobalStatisticsAggregator<K> lastCompletedAggregator;
+
+  GlobalStatisticsAggregatorTracker(DataStatisticsFactory<K> statisticsFactory, int parallelism) {
+    this.statisticsFactory = statisticsFactory;
+    this.parallelism = parallelism;
+  }
+
+  boolean receiveDataStatisticEventAndCheckCompletion(int subtask, DataStatisticsEvent<K> event) {
+    LOG.info("!!! Receive event {} from subtask {}", event, subtask);
+    long checkpointId = event.checkpointId();
+
+    if (lastCompletedAggregator != null && lastCompletedAggregator.checkpointId() >= checkpointId) {
+      LOG.debug(
+          "Data statistics aggregation for checkpoint {} has completed. Ignore the event from subtask {} for checkpoint {}",
+          lastCompletedAggregator.checkpointId(),
+          subtask,
+          checkpointId);
+      return false;
+    }
+
+    if (inProgressAggregator == null) {
+      inProgressAggregator = new GlobalStatisticsAggregator<>(checkpointId, statisticsFactory);
+    }
+
+    if (inProgressAggregator.checkpointId() < checkpointId) {
+      if ((double) inProgressAggregator.aggregatedSubtasksCount() / parallelism * 100
+          >= EXPECTED_DATA_STATISTICS_RECEIVED_PERCENTAGE) {
+        lastCompletedAggregator = inProgressAggregator;
+        LOG.info(
+            "Received data statistics from {} operators out of total {} for checkpoint {}. "
+                + "It's more than the expected percentage {}. Update last completed aggregator to be "
+                + " {}.",
+            inProgressAggregator.aggregatedSubtasksCount(),
+            parallelism,
+            inProgressAggregator.checkpointId(),
+            EXPECTED_DATA_STATISTICS_RECEIVED_PERCENTAGE,
+            lastCompletedAggregator);
+        inProgressAggregator = new GlobalStatisticsAggregator<>(checkpointId, statisticsFactory);
+        inProgressAggregator.mergeDataStatistic(subtask, event);
+        return true;
+        //        sendDataStatisticsToSubtasks(
```

Comment:
Nit: remove?

Reply:
I forget to remove it. Thanks for catching it. I will re